osquery/tables/system/windows/objects.cpp

```diff
@@ -0,0 +1,229 @@
+// "\Sessions\BNOLINKS\<sessionnum>"
+//
+Status enumerateBaseNamedObjectsLinks(const std::wstring& session_num,
+                                      const std::wstring& object_type,
+                                      std::vector<obj_name_type_pair>& objects) {
+  // look up NtQuerySymbolicLinkObject as exported from ntdll
+  //
+  auto NtQuerySymbolicLinkObject =
+      (NTQUERYSYMBOLICLINKOBJECT)GetProcAddress(GetModuleHandleA("ntdll"),
+                                                "NtQuerySymbolicLinkObject");
+  if (NULL == NtQuerySymbolicLinkObject) {
+    return Status(GetLastError(), "Cannot locate NtQuerySymbolicLinkObject");
+  }
+
+  // by convention, we expect there to be <n> objects in \Sessions\BNOLINKS with
+  // the following three characteristics:
+  //
+  //   (1) the object name is the string representation of of an active terminal
+  //       services session id.  this means we expect the object name to be a
+  //       string representation of an integer
+  //
+  //   (2) the object type name be "SymbolicLink"
+  //
+  //   (3) the symbolic link point to a directory object
+  //
+  // validate in this order
+  //
+
+  // validate (1)
+  //
+  // validate that this appears to be a valid terminal services session id
+  // another approach is to enumerate windows terminal services sessions with
+  // WTSEnumerateSessions and validate against that list
+  //
+  if (!(L"0" == session_num || safeWstrToInt(session_num) > 0)) {
+    return Status(ERROR_INVALID_PARAMETER, "Unrecognized Session Id");
+  }
+
+  // validate (2)
+  //
+  // validate that the object type name is "SymbolicLink"
+  //
+  if (L"SymbolicLink" != object_type) {
+    return Status(ERROR_INVALID_PARAMETER, "Unexpected Object Type");
+  }
+
+  // at this point we have SymbolicLink with a name matching a terminal services
+  // session id.   build the fully qualified object path
+  //
+  std::wstring qualifiedpath = L"\\Sessions\\BNOLINKS\\" + session_num;
+
+  // open the symbolic link itself in order to determine the target of the link
+  //
+  Handle slo;
+  auto status = slo.openSymLinkObj(qualifiedpath);
+  if (!status.ok()) {
+    // log?
+    return status;
+  }
+
+  UNICODE_STRING usSymbolicLinkTarget;
+  WCHAR wzTargetLinkBuffer[MAX_PATH];
+  usSymbolicLinkTarget.Buffer = wzTargetLinkBuffer;
+  usSymbolicLinkTarget.Length = 0;
+  usSymbolicLinkTarget.MaximumLength = MAX_PATH;
+
+  auto ntStatus =
+      NtQuerySymbolicLinkObject(slo.getAsHandle(), &usSymbolicLinkTarget, NULL);
+  if (STATUS_SUCCESS != ntStatus) {
+    return Status(ntStatus, "NtQuerySymbolicLink failed");
+  }
+
+  return enumerateObjectNamespace(usSymbolicLinkTarget.Buffer, objects);
+}
+
+QueryData genBaseNamedObjects(QueryContext& context) {
+  QueryData results;
+
+  // enumerate the base named objects in each terminal services session
+  //
+  std::vector<obj_name_type_pair> sessions;
+  auto status = enumerateObjectNamespace(L"\\Sessions\\BNOLINKS", sessions);
```

Comment:
Looks like we use `L"\\Sessions\\BNOLINKS"` more than once, maybe make it a const global? `const std::wstring kBnoLinks { L"\\Sessions\\BNOLINKS" };`?

Reply:
@muffins ok great will do

Reply:
done